### PR TITLE
Fix `!=` operator for MySQL

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Fix the `!=` operator for MySQL.

### Comparison of SQL generated by `case_sensitive_mysql_ne`:

`filter_string="tag.key1 != 'value'"` generates a SQL like this:

Before

```
WHERE experiment_tags.key = %s AND BINARY experiment_tags.key = %s AND experiment_tags.value != %s OR BINARY experiment_tags.value != %s
```

After

```
WHERE experiment_tags.key = %s AND BINARY experiment_tags.key = %s AND (experiment_tags.value != %s OR BINARY experiment_tags.value != %s)
```

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 1]
```diff
@@ -410,29 +410,31 @@ def test_search_experiments_filter_by_attribute(self):
...
+        experiments = self.store.search_experiments(filter_string="tag.key1 = 'value'")
         assert [e.name for e in experiments] == ["exp1"]
-        experiments = self.store.search_experiments(filter_string="tag.`k e y` = 'value'")
+        experiments = self.store.search_experiments(filter_string="tag.`k e y 1` = 'value'")
         assert [e.name for e in experiments] == ["exp3"]
-        experiments = self.store.search_experiments(filter_string="tag.\"k e y\" = 'value'")
+        experiments = self.store.search_experiments(filter_string="tag.\"k e y 1\" = 'value'")
         assert [e.name for e in experiments] == ["exp3"]
-        experiments = self.store.search_experiments(filter_string="tag.key != 'value'")
+        experiments = self.store.search_experiments(filter_string="tag.key1 != 'value'")
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:426]: Prior to the fix, this line failed with the following error:

```
        experiments = self.store.search_experiments(filter_string="tag.key1 != 'value'")
>       assert [e.name for e in experiments] == ["exp2"]
E       AssertionError: assert ['exp2', 'exp2'] == ['exp2']
E         Left contains one more item: 'exp2'
E         Full diff:
E         - ['exp2']
E         + ['exp2', 'exp2']
```

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
I have a follow-up question,

for filters constructed by `sa.and_`, / `sa.or_`, does it have similar issue ?

e.g.

`sa.and_(sa.or_(f1, f2), sa.or_(f3, f4))`

does it generate `(f1 OR f2) AND ( f3 OR f4 )` ?

**Comment by harupy** [comment]
@WeichenXu123 

> sa.and_, / sa.or_, does it have similar issue ?

Looks like it does. See https://stackoverflow.com/questions/13370993/sqlalchemy-query-and-or-issue.

**Comment by WeichenXu123** [comment]
What about use `.self_group` instead of adding `()` ? And for safety let's add it to all `case_sensitive_mysql_*` expressions .

**Comment by harupy** [comment]
I tried `self_group` but it didn't add `()`:

```diff
diff --git a/mlflow/utils/search_utils.py b/mlflow/utils/search_utils.py
index f77a596ed..4918fb0df 100644
--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -115,8 +115,10 @@ class SearchUtils:
             )
 
         def case_sensitive_mysql_ne(value):
-            return sa.text(f"({col} != :value OR BINARY {col} != :value)").bindparams(
-                sa.bindparam("value", value=value, unique=True)
+            return (
+                sa.text(f"{col} != :value OR BINARY {col} != :value")
+                .bindparams(sa.bindparam("value", value=value, unique=True))
+                .self_group()
             )
 
         def case_sensitive_mysql_like(value):
```

**Comment by harupy** [comment]
`TextClause.self_group`'s implementation:

https://github.com/zzzeek/sqlalchemy/blob/fc5c54fcd4d868c2a4c7ac19668d72f506fe821e/lib/sqlalchemy/sql/elements.py#L2169-L2175

```
    def self_group(self, against=None):
        if against is operators.in_op:
            return Grouping(self)
        else:
            return self
```

**Comment by WeichenXu123** [comment]
> @WeichenXu123
> 
> > sa.and_, / sa.or_, does it have similar issue ?
> 
> Looks like it does. See https://stackoverflow.com/questions/13370993/sqlalchemy-query-and-or-issue.

I just tested, it can automatically add () correctly in this case:

```
select(SqlModelVersion).filter(
            sqlalchemy.and_(
                sqlalchemy.or_(SqlModelVersion.name == 'a', SqlModelVersion.name == 'b'),
                sqlalchemy.or_(SqlModelVersion.version == '1', SqlModelVersion.version == '2'),
            )
        )
```
generates:
```
SELECT model_versions.name, model_versions.version, model_versions.creation_time, model_versions.last_updated_time, model_versions.description, model_versions.user_id, model_versions.current_stage, model_versions.source, model_versions.run_id, model_versions.run_link, model_versions.status, model_versions.status_message \nFROM model_versions \nWHERE (model_versions.name = ? OR model_versions.name = ?) AND (model_versions.version = ? OR model_versions.version = ?)
```

and 
```
select(SqlModelVersion).filter(
            sqlalchemy.or_(
                sqlalchemy.and_(SqlModelVersion.name == 'a', SqlModelVersion.name == 'b'),
                sqlalchemy.and_(SqlModelVersion.version == '1', SqlModelVersion.version == '2'),
            )
        )
```
generates:
```
SELECT model_versions.name, model_versions.version, model_versions.creation_time, model_versions.last_updated_time, model_versions.description, model_versions.user_id, model_versions.current_stage, model_versions.source, model_versions.run_id, model_versions.run_link, model_versions.status, model_versions.status_message \nFROM model_versions \nWHERE model_versions.name = ? AND model_versions.name = ? OR model_versions.version = ? AND model_versions.version = ?
```

**Comment by harupy** [comment]
@WeichenXu123 

> I just tested, it can automatically add () correctly in this case:

Are you sure? The generated query doesn't contain `()` at all.

**Comment by harupy** [comment]
Is `a AND b OR c AND d` equivalent to `(a AND b) OR (c AND d)`?

**Pull request merged**